### ISSUE TITLE
CognitoCachingAnonymousCredentialsProvider::GetCredentialsFromCognito() unnecessarily needs account id

### ISSUE THREAD

**Issue by orinem** [aws/aws-sdk-cpp#1448]
Confirm by changing [ ] to [x] below to ensure that it's a bug:
- [d ] I've gone though [Developer Guide](https://docs.aws.amazon.com/sdk-for-cpp/v1/developer-guide/welcome.html) and [API reference](http://sdk.amazonaws.com/cpp/api/LATEST/index.html)
- [ d] I've searched for [previous similar issues](https://github.com/aws/aws-sdk-cpp/issues) and didn't find any solution
  
**Describe the bug**

**_NOTE: this does not happen for the same authentication flow with the .net sdk._**

Attempt to use classes derived from Aws::Auth::PersistentCognitoIdentityProvider and Aws::Auth::CognitoCachingAuthenticatedCredentialsProvider without an account id.

There is no way to specify no account id.

FetchCredentialsFromCognito() in CognitoCachingCredentialsProvider.cpp _always_ calls getIdRequest.SetAccountId(accountId);

If accountId is empty, the GetId() call will fail.

Note that AccountId on the underlying REST API is NOT required:

https://docs.aws.amazon.com/cognitoidentity/latest/APIReference/API_GetId.html

**SDK version number**

Latest, i.e. 1.8.39

**Platform/OS/Hardware/Device**

Not particularly relevant, but Windows 10

**To Reproduce (observed behavior)**

See above.

**Expected behavior**

The c++ SDK should handle the situation the same as the .net SDK.  I.E. there should be a way of specifying no account id.  In this case, the .net SDK checks for a null account id.  Since the c++ SDK uses Aws::String, null is not an option.  However, since the specification for the AccountId in the REST API is "\d+", an empty string would be sufficient to indicate no account id.

In other words, replace:

```
        GetIdRequest getIdRequest;
        getIdRequest.SetAccountId(accountId);

```
with:
```
		GetIdRequest getIdRequest;
		if (!accountId.empty())
		{
			getIdRequest.SetAccountId(accountId);
		}
```

There is a similar argument to be made for the logins parameter where again, the .net SDK checks the length of the logins array and does not marshal the array if empty.

Fortunately, it is possible to override the GetCredentialsFromCognito() function from CognitoCachingAuthenticatedCredentialsProvider and implement the above change.  However, that is not a good long term solution should anything else in CognitoCachingAuthenticatedCredentialsProvider change.

**Comment by robertmchardy** [comment]
We are having the same issue with providing an (unnecessary) accountId to the credentials provider - not ideal to distribute account id in our C++ apps, and different from all other SDKs.

Is there any progress on getting this into a new SDK release?

Thanks, Rob.

**Comment by KaibaLopez** [comment]
Correct me if I'm missing something , but the call does require an Identity Id, all that code is trying to do is finding it when it is not provided and using account id to do so. So it's not that accountId is required but that it can be used to fetch the identityId. 
So in short, if both the identityId and account Id are both missing, the call should fail right?

**Comment by robertmchardy** [comment]
> Correct me if I'm missing something , but the call does require an Identity Id, all that code is trying to do is finding it when it is not provided and using account id to do so. So it's not that accountId is required but that it can be used to fetch the identityId.
> So in short, if both the identityId and account Id are both missing, the call should fail right?

Yes, that's correct, but not the issue here.  We were finding that even when the identity id was provided, that the **account id** had also to be provided (unnecessarily), and would fail if not.

**Comment by NworbLegin** [comment]
I concur with @robertmchardy that this is an unnecessary parameter for this use case.

**Comment by KaibaLopez** [comment]
Ok sorry for the confusion I do see the problem.
The root cause is on the cognito's service API . As you say, the getId call will fail if an unacceptable AccountId is part of it, even if the idetity-pool-id is provided, but this happens from the API level.

Having said that, yea the workaround seems valid and way faster than trying to get the service team to fix the model so I'll give it a try. 
Thanks for clarifying.

**Comment by lmbr-pip** [comment]
Looks like this may have been fixed in https://github.com/aws/aws-sdk-cpp/commit/b21be81543755fe2d28f4f9d14800992c01ca583

**Comment by KaibaLopez** [comment]
yea, the proposed solution has been added to the SDK, thanks for the info and push to getting this done!
